hibernate-core/src/main/java/org/hibernate/type/descriptor/java/GenericArrayTypeDescriptor.java

```diff
@@ -0,0 +1,358 @@
+		this( baseDescriptor, ObjectSqlTypeDescriptor.INSTANCE );
+	}
+
+	public GenericArrayTypeDescriptor(JavaTypeDescriptor<T> baseDescriptor, SqlTypeDescriptor sqlTypeDescriptor) {
+		super(
+				(Class<T[]>) Array.newInstance( baseDescriptor.getJavaTypeClass(), 0 ).getClass(),
+				determineMutabilityPlan( baseDescriptor )
+		);
+		this.componentDescriptor = baseDescriptor;
+		this.componentClass = baseDescriptor.getJavaTypeClass();
+		this.componentSqlDescriptor = sqlTypeDescriptor;
+	}
+
+	private static <T> MutabilityPlan<T[]> determineMutabilityPlan(JavaTypeDescriptor<T> baseDescriptor) {
+		return new LocalArrayMutabilityPlan<>( baseDescriptor );
+	}
+
+	@Override
+	public JavaTypeDescriptor<T> getElementDescriptor() {
+		return componentDescriptor;
+	}
+
+	@Override
+	public BasicType<?> createType(BasicType<T> elementType) {
+		final Class<?> elementJavaTypeClass = elementType.getJavaTypeDescriptor().getJavaTypeClass();
+		if ( elementType instanceof BasicContainerType<?> || elementJavaTypeClass != null && elementJavaTypeClass.isArray() ) {
+			return null;
+		}
+		if ( componentSqlDescriptor != elementType.getSqlTypeDescriptor() ) {
+			return new BasicArrayType<>( elementType, new GenericArrayTypeDescriptor<>( elementType ) );
+		}
+		return new BasicArrayType<>( elementType, this );
+	}
+
+	@Override
+	public String extractLoggableRepresentation(T[] value) {
+		return value == null ? super.extractLoggableRepresentation( null ) : Arrays.toString( value );
+	}
+
+	private static class LocalArrayMutabilityPlan<T> implements MutabilityPlan<T[]> {
+
+		private final Class<T> componentClass;
+		private final MutabilityPlan<T> componentPlan;
+
+		public LocalArrayMutabilityPlan(JavaTypeDescriptor<T> baseDescriptor) {
+			this.componentClass = baseDescriptor.getJavaTypeClass();
+			this.componentPlan = baseDescriptor.getMutabilityPlan();
+		}
+
+		@Override
+		public boolean isMutable() {
+			return true;
+		}
+
+		@Override
+		public T[] deepCopy(T[] value) {
+			if ( value == null ) {
+				return null;
+			}
+			T[] copy = (T[]) Array.newInstance( componentClass, value.length );
+			for ( int i = 0; i < value.length; i ++ ) {
+				copy[ i ] = componentPlan.deepCopy( value[ i ] );
+			}
+			return copy;
+		}
+
+		@Override
+		public Serializable disassemble(T[] value) {
+			return (Serializable) deepCopy( value );
+		}
+
+		@Override
+		public T[] assemble(Serializable cached) {
+			return deepCopy( (T[]) cached );
+		}
+
+	}
+
+	@Override
+	public boolean areEqual(T[] one, T[] another) {
+		if (one == null && another == null) {
+			return true;
+		}
+		if (one == null || another == null) {
+			return false;
+		}
+		if (one.length != another.length) {
+			return false;
+		}
+		int l = one.length;
+		for (int i = 0; i < l; i++) {
+
+			if ( ! componentDescriptor.areEqual( one[i], another[i] )) {
+				return false;
+			}
+		}
+		return true;
+	}
+
+	@Override
+	public int extractHashCode(T[] value) {
+		return java.util.Arrays.hashCode( value );
```

Comment:
If we implement 'areEqual()' using componentDescriptor's 'areEqual()' internally, we might well use its 'extractHashCode' in the above 'extractHashCode()' implementation as well; otherwise there would be the danger that the contract between 'areEqual()' and 'extractHashCode()' might not be satisfied. A possible downside is we can insert some instance into a Map or Set but we won't be able to retrieve it.
Not 100% sure though.